Add User Guide (#175)
k8s-service-bindings/spec: docs/user-guide.md

### Purpose of the type and the provider fields in the application projection

Service Binding specification mandates `type` field and recommends `provider` field in the binding Secret resource.  The provider field is recommended to support scenarios where there could be different providers for the same Provisioned Service type.  For example, if the type is `mysql`, the provider value could be `mariadb`, `oracle`, `bitnami`, `aws-rds`, etc.  When the application is reading the binding values, if necessary, the application could consider `type` and `provider` as a composite key to avoid ambiguity.  It will be helpful if an application needs to choose a particular provider based on the deployment environment.  In the deployment environment (`stage`, `prod`, etc.), at any given time, you need to ensure only one bindings projection exits for a given `type` or `type` & `provider` -- unless your application needs to connect to all the services.

> The operator autogenerates the binding directory name, and relying on the binding name is fragile.  So, it is not a good practice to use the binding name to look up the bindings.

Reviewer: I thought we agreed yesterday that the operator should not generate the binding name. The bit about encouraging users to primarily bind by type is good, but there are times you need to distinguish two equivalent types by name

Author: Sorry, I missed it. I will change this text.

### Programming language specific library APIs

The application can retrieve the bindings through the library available for your programming language of choice.  The language-specific APIs are encouraged to follow the pattern described here.  This is not an exhaustive list of APIs.  Consult your library API documentation to confirm the behavior.

> The language binding API can create separate functions to retrieve bindings. For example, in Go:

Reviewer 2: ```suggestion
For languages such as Go without operator overloading, separate functions can be used to retrieve bindings:
```

```

> Bindings(_type string) []map[string]string

Reviewer 2: ```suggestion
Bindings(_type string) []Binding
```

> BindingsWithProvider(_type, provider string) []map[string]string

Reviewer 2: ```suggestion
BindingsWithProvider(_type, provider string) []Binding
```

```

> Languages with function overloading can use a method with different parameters to retrieve bindings. For example, in Java:

Reviewer 2: ```suggestion
For languages such as Java with operator overloading, the same method name with different argument lists can be used to retrieve bindings:
```

```
public List<Binding> filterBindings(@Nullable String type)
public List<Binding> filterBindings(@Nullable String type, @Nullable String provider)
```

(Example taken from [Spring Cloud Bindings](https://github.com/spring-cloud/spring-cloud-bindings))

> Since the APIs are returning more than one value, depending on your application need, you can choose to connect to the first entry or all of them, if required.

Reviewer 2: ```suggestion
The specification does not guarantee a single binding of a given type or type, provider tuple so the APIs return collections of bindings.  Depending on your workload need, you can choose to connect to the first entry or all of them.
```

### Environment Variables

> The spec also has support for projecting environment variables.  You can use the built-in language feature of your programming language of choice to read environment variables.  The container must restart to update the values of environment variables if there is a change in the Secret resource.

Reviewer 2: ```suggestion
The specification also has support for projecting binding values as environment variables.  You can use the built-in language feature of your programming language of choice to read environment variables.  The container must restart to update the values of environment variables if there is a change in the binding.
```

Reviewer 2: Much like the list name collision above, I feel like this should be a failure-to-reconcile if env-var names collide.  /cc @arthurdm and @scothis
